java/vector/src/main/java/org/apache/arrow/vector/compare/RangeEqualsVisitor.java

```diff
@@ -68,7 +67,13 @@ private void checkType() {
     } else if (left == right) {
       typeCompareResult = true;
     } else {
-      typeCompareResult = left.getField().getType().equals(right.getField().getType());
+      if (left instanceof UnionVector || right instanceof UnionVector) {
+        // for UnionVectors, it is possible that two vectors can have different
```

Comment:
by this argument, wouldn't we need to always return true here and validate types in equalities.
i.e. if left is a union and right is an intvector   it union[1] happens to be an int vector the comparison should happen.

Reply:
I think the argument only applies when both vectors are union vectors. 
An IntVector and a UnionVector can not have equal range, even the values in the range are actually equal.

In C++, it also requires that both vectors are union vectors:

bool CompareUnions(const UnionArray& left) const {
    const auto& right = checked_cast<const UnionArray&>(right_);

Reply:
its been a while since I looked at the C++ code, but are you sure there isn't a check for union equality first in C++?  It doesn't seem unreasonable to require the types be identical.

Reply:
There is a check for union type euqlity (and other types of arrays) in C++, and it requres the type to be identical. 

To see this, note in the above code snippet that the left vector must be a UnionArray (passed in as the parameter), and the right vector must also be a UnionArray, and this is ensured by 

checked_cast<const UnionArray&>(right_);

Reply:
I think maybe I was not answering your question, and there is something confusing here. So let me try to make it clearer:

There are two types of type equality: 
shallow equality - which only requires the minor types to be the same 
deep equality - which requires the minor type, minor types of child types, and order of child types be equal. 

For Java, we check deep equality for other types of vectors, and for union vector, we only check shallow equality. 

This is because, for union vector, it is possible that two union vectors may have different child vectors, but their values are equal in a certain range. This is not possible for other types of vectors.

Reply:
I'm not really sure this would be considered a bug.  I don't think there is anything wrong to say for union vectors the child vectors have to be exactly the same before comparing values over a particular range.

Reply:
Logically, a union vector does not have any child vectors. It just has a number of heterogeneous value slots. So child vectors are just implementation specific issues, which has not much to do with the union vector specification.

IMO, by comparing a range of values, we only need to make sure the slots within the range are equal (type equal + value equal)

That is, the equal range implementation should be consistent with the following specification:

for (int i = range.start; i < range.end; i++) {
  left.get(i).equals(right.get(i));
}

It can be seen that the above specification is consistent with the C++ implementation:

    if (union_mode == UnionMode::SPARSE) {
        if (!left.child(child_num)->RangeEquals(i, i + 1, o_i, right.child(child_num))) {
          return false;
        }
      } else {
        const int32_t offset = left.raw_value_offsets()[i];
        const int32_t o_offset = right.raw_value_offsets()[o_i];
        if (!left.child(child_num)->RangeEquals(offset, offset + 1, o_offset,
                                                right.child(child_num))) {
          return false;
        }
      }

Note that a single slot is compared each time.

Reply:
The code you pasted is the fixed compareUnionVectors, which seems like a correct fix here.  The change to type comparison here, seems more arbitrary.  The semantics aren't really well defined.  Which goes back to my comment above.  Any of the following could be valid (which is why I don't see a reason to change it):
1.  It is OK for unions to compare against any other vector as long as the range is valid.
2.  It is ok for unions to compare against any other type of union (dense or sparse)
3.  It is only ok to compare unions that have the exact same metadata.

The C++ implementation seems to, assume that type codes are equivalent for both unions which they wouldn't have to be in practice (i.e they are part of the metadata).  Also I think the ArrayRangeEquals in C++ has a bug in that is only checking type codes (ArrayEquals does check for type equality).

Reply:
@emkornfield I agree with you that there can be more than one valid semantics for range equality. I also agree that the above 3 semantics are valid to some extent. 

The semantics for the current code performs the comparison in two steps:
1. compare type codes of child vectors.
2. compare values slots in the specified range.

I can be argued that this is also a valid range equality semantics. However, I am not sure if it is useful and practical. Maybe it will suprise the users. 

For example, when encoding a union vector, the union vector contains child vectors for all dictionary entries. However, the vector being encoded may not contain all child vectors. Therefore, the vector to encode cannot find a match in the dictionary, even if the values in the range are equal.

Reply:
> compare type codes of child vectors.
This is what I'm think is a subjective change.  The old way of requiring metadata for UnionVectors be equals seems valid.

> For example, when encoding a union vector, the union vector contains child vectors for all dictionary entries. However, the vector being encoded may not contain all child vectors. Therefore, the vector to encode cannot find a match in the dictionary, even if the values in the range are equal.

I don't understand this, can you provide an example of the problem you are highlighting?  Dictionary encoding in general can use some improvement in the java code.

Reply:
Sure. Let me try to provide a simple example.
Suppose our dictionary is a union vector: {Int (0), Long(1), Byte(2)}. This vector has 3 child vectors: IntVector, BigIntVector, and SmallIntVector.

We want to encode another union vector: {Int(0), Byte(2)}. The encoded vector should be an integer vector {0, 2}. 

However, since the vector to encode has only 2 children: IntVector and SmallIntVector, the rangeEqualVisitor will not pass the type check (according to the current code base), and range check will always return false, and no values will be matched in the dictionary.

Reply:
I don't think this is an issue.  This is technically using a different dictionary type than the given vector that is trying to be encoded, which I think is ok to disallow.

Reply:
Thanks for your comments. I am not sure if I fully understand your point. Please help point out what is wrong with my reasoning below. 

In the example above, the dictionary is a UnionVector, and the vector to encode is also a UnionVector. So they have the same type?

You may regard them as of different types, because they have different child vectors. The fundamental reason for this is that the vector to encode only contains a proper subset of values in the dictionary. 

I think this is a common scenario (the vector to encode is a proper subset of the dictionary), which is frequently encountered in practice and should be supported. 

To see this, note that a dictionary may be used to encode many vectors (maybe the dataset is split into many vectors with manageable size), so each vector contains only a small part of the dataset, and it is likely that the elements in a vector is only a proper subset of the dictionary. 

If we cannot reach a conclusion, do you think I should start a discussion in the ML?

Reply:
If it were a StructVector, it would be natural to consider them as different types because of different child vectors. For a UnionVector, they should be considered as of the same type, because child vectors are interpreted in a different way. 

All UnionVectors should be considered as having the same type (this is a special feature for UnionVector, which is not true for StructVector, ListVector, etc.), because each value slot can have any possible type of value. 

If this problem is not resloved, dictionary encoding for UnionVector will not work properly.

Reply:
I'm not saying the interpretation proposed here is wrong, just that there are multiple interpretations and the current one is valid.  As noted above I don't see why any other type can't be encoded in using a Union dictionary given the reasoning about subranges being equal.  This is also not to say that there can't be other visitors that have different equality semantics, but I don't think there is a strong enough justification to change the current one.  I'm also not clear that this presents a real problem.  Feel free to start a thread on the mailing list, as there are multiple interpretations so other might feel strongly about some of the other options.

In addition, the current the current Java implementation isn't as general as the spec allows since it doesn't allow children vectors of the same primitive type (which the specification does), and if this feature was used, I would argue the equality of metadata on children vector types becomes important (this has been a matter of previous debate that hasn't been fully resolved either).
> Sparse and Dense Union: a nested layout representing a sequence of values, each of which can have type chosen from a collection of child array types.

Reply:
Thanks for your comments. I think some points are a little clearer for me. 

I think there are two (somewhat) independent problems here:
1. Should child vector types being equal be a necessary condition for vector ranges being equal. 
2. Should we consider the above 3 (or other alternative) semantics proposed by you.

For me, 1 is unacceptable, as it causes practical problems (e.g. encoding a UnionVector). For 2, I do not prefer it, but it is acceptable for me.

Reply:
I'm not sure they are independent.  But at this point, if you still feel strongly about 1, please start a discussion on the mailing list.

Reply:
Sure. Thanks for your suggestion. I will start a discussion in the ML.

Reply:
It doesn't seem like there was much discussion on this in the mailing list, would you be willing to separate the fixes here, so we can merge the bug fix and maybe try bumping the thread on the mailing list?

Reply:
Sounds like a good idea. I have revised the PR to undo the union type comparison logic. 
It seems like a seprate issue, and may involve multiple languages. 

Maybe I will do more investigations and start a new discussion to describe the problem in a clearer way. 

This discussion have been for a long time. Thanks a lot for your patience.